README
About this distro
-----------------

This collection of scripts and modules provides access to the APC
(Archive of Perl Changes).

Data Prerequisites
------------------

You need to mirror parts of the APC repository to your local disk.
This is described in the buildaperl manpage.

Scripts
-------

patchaperlup                 a low level script that only builds
                             new perl sources from old perl
                             sources by applying numerically
                             sorted patchfiles.

buildaperl                   a wrapper around patchaperlup that
                             is able to compile and install a new
                             perl into its own cache area.

binsearchaperl               a wrapper around buildaperl that is
                             able to run already built perls from the
                             cache area as well as build new perls via
                             buildaperl and compare their
                             functionality. If there is a different
                             behaviour between two perls, the script
                             is able to determine the exact patch
                             number that caused the change.

perlpatch2svn                *** DEPRECATED ***
                             parse a stream of patches from the
                             APC and feed them into a Subversion
                             repository. Pedantically reconstruct
                             as much as possible from the
                             Perforce repository, binary files
                             that are missing in the APC are
                             restored from a dump.

apc2svn                      *** DEPRECATED ***
                             a wrapper around perlpatch2svn which
                             has a concept how a future
                             subversion repository might look
                             like. It is able to import the whole
                             APC into an empty or prepopulated
                             repository. Incremental updates from
                             APC into this repository supported.
                             **Use with caution**, the repository
                             layout itself is still subject to
                             change.


Modules
-------

Perl::Repository::APC        provides methods to get at the
                             important metadata of the
                             APC--branches, perl versions and
                             patches.

Perl::Repository::APC2SVN    *** DEPRECATED ***
                             a collection of utility functions
                             for the scripts that deal with
                             importing the APC into a Subversion
                             repository.

Perl::Repository::APC::BAP   Rules how to translate arguments
                             5.6.0@5666, 5.7.3@, @17000, etc. as
                             short names for patched perl
                             versions. Only used by buildaperl.

Copyright
---------

Code and data are copyright (C) 2002,2003,2005,2006,2007,2008 Andreas
Koenig and Rafael Garcia-Suarez. They can be used, modified, and
redistributed under the same terms as Perl itself.


Changes
2008-11-28  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* Release 2.002:

	Workaround for broken patch 34882: supply the whole file
	t/op/regexp_unicode_prop_thr.t from within the script.

2008-03-08  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* Release 2.001:

	Unchanged stable release. For the changes between the 1.276 release and
	this release, read below.

2008-02-22  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* Release around 2.000_294:

	* Adding support for 5.10.1. Tiny improvements to code readability by
	adding labels and changing bad variable names.

2008-01-25  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* Release around 2.000_284:

	* The APC has got a new directory with additional tarballs (thanks to
	Phillippe Chiasson) so we can drop the README.additional_tarballs and
	simplify the documentation.

	Leaving the underscore in the version number to test the package for a
	while.

2008-01-20  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* Release around 2.000_280:

	* Support for additional tarballs that are currently not in the APC. See
	README.additional_tarballs for instructions how to enable the new
	feature by copying your favorite tarballs to a new selfmade directory.

	With this release we support 15 more tarballs than those in the APC,
	which results in a muchly reduced time for patching.

	The internal datastructures have changed, so we make this a dev release
	and inc the major version number. In principle all changes should be
	backwards compatible: P:R:BAP->translate returns a fifth element.
	P:R:APC->tarball looks into more directories than it did before and may
	return an absolute path, no other changes in the modules are visible to
	the scripts.

2008-01-03  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* Release around 1.275:

	* bugfix: address https://rt.cpan.org/Ticket/Display.html?id=32052 --
	set LD_LIBRARY_PATH when calling ./perl.

2007-11-11  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* Release around 1.272:

	* bugfix: check the exit status of prep scripts and if they fail rerun
	them up to two times. That prevents bad results in scenarios with
	complex upgrade scripts.

2007-09-09  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* trimtrees now ignores empty files

2007-08-12  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* Release around 1.258:

	* new switch --writedotpatch to create a .patch file that governs
	PERL_PATCHLEVEL in the standard perl build procedure

2007-07-10  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* Release around 1.249:

	* applied a patch from Father Chrysostomos to integrate better into OS X
	environment

	* a few tiny cosmetic nits not worth mentioning

	* I wanted to look into CPANTS if there were any issues open but they
	were closed for maintainance

2007-04-24  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* Release around 1.243:

	* remove empty directories immediately because the APC patch files do
	not contain directory removal information, as patch 31047 painfully
	demonstrated.

	* document under BUGS that patchaperlup does not work on case tolerant
	filesystems.

2007-02-14  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* Release around 1.239:

	* perl 5.6.2 failed one test due to different walk through find(). Only
	the badly constructed test needed fixing

2007-02-12  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* Release around 1.236:

	* add prereq Digest::MD5

2007-02-10  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* Release around 1.233 features

	* the tempdirs that end up in the chache are now mode 0755

	* do not ask back if a *.t file is going to be deleted (this was just a
	misfeature)

	* improve the concurrency protection so that many binsearches can run in
	parallel

2005-11-30  Andreas J. Koenig  <andreas.koenig.gmwojprw@franz.ak.mind.de>

	* Today's release (1.169 or so) features
 
	* several bugfixes and enhancements in eg/trimtrees.pl (thanks to Eric
	Schnetter for reprting bugs and testing the fixes).

	

	Local Variables:
	mode: change-log
	change-log-default-name: "Changes"
	tab-width: 2
	left-margin: 2
	End:


eg/trimtrees.pl
#!/usr/bin/perl -w

use strict;
use Digest::MD5 qw(md5);
use File::Find;
use File::Compare qw(compare);
use Cwd;

my @dirs = @ARGV or die "Usage: $0 directories";
our $Signal = 0;
$SIG{INT} = sub {
  warn "Caught SIGINT; please stand by, I'm leaving as soon as possible...\n";
  $Signal++;
};

sub fmt ($) {
  local $_ = shift;
  s/(\d)(?=(\d{3})+$)/$1_/g;
  $_;
}

undef $/;
our %MD5;
my %reported;
my $files = 0;
my $dirs = @dirs;
my $savedspc = 0;
my $usedspc = 0;
my $tl_dirs_todo = 0;
my $tl_dirs_doing = 0;
my $WD = Cwd::cwd;
my $sprintf ="\rtlds[%s]doing[%s]uniqfils[%s]fils[%s]usedspc[%s]savdspc[%s]";

$| = 1;
for my $diri (0..$#dirs) {
  my $root = $dirs[$diri];
  find(
       {
        wanted => sub {
          if ($Signal){
            $File::Find::prune = 1;
            return;
          }
          if ($File::Find::name eq $root) {
            my $td = $_;
            opendir my($dh), $td;
            my(@tl) = grep { !/^\./ && -d "$td/$_" } readdir $dh;
            $tl_dirs_todo = @tl;
          } elsif (-d) {
            my $slashes = $File::Find::name =~ tr|/||;
            if ($slashes == 1) {
              $tl_dirs_doing++;
            }
          }
          return if -l; # relative links would need special treatment that does not pay off
          return unless -f _;
          my $basename = $_;
          open my $fh, "<", $basename;
          my $data = <$fh>;
          my $md5 = md5 $data;
          $files++;
          my $cand = $File::Find::name;
          if (my $first = $MD5{$md5}) {
            $first = File::Spec->file_name_is_absolute($first) ?
                $first : File::Spec->catfile($WD, $first);
            my(@refstat) = stat($first);
            die "illegal refstat[@refstat]first[$first]" unless $refstat[1];
            die "Sensation, $first and $cand are not equal with same MD5"
                if compare $first, $basename;
            my(@candstat) = stat($basename);
            return unless $candstat[0] == $refstat[0]; # different file system
            return if $candstat[1] == $refstat[1];     # already same inode
            if (unlink $basename and link $first, $basename) {
              $savedspc += $refstat[7];
            } else {
              die "ERROR: first[$first]cand[$cand]![$!]";
            }
          } else {
            $MD5{$md5} = $cand;
            my $size =  -s $cand;
            warn "size undefined" unless defined $size;
            warn "usedspc undefined" unless defined $usedspc;
            $usedspc += $size;
          }
          my $uniq_files = keys %MD5;
          return if ($uniq_files % 100 || $reported{$uniq_files}++);
          printf(
                 $sprintf,
                 map { fmt($_) }
                 $tl_dirs_todo,
                 $tl_dirs_doing,
                 $uniq_files,
                 $files,
                 $usedspc,
                 $savedspc
                );
        },
        no_chdir => 1,
       },
       $root
      );
  last if $Signal;
}
my $uniq_files = keys %MD5;
printf(
       $sprintf,
       map { fmt($_) }
       $tl_dirs_todo,
       $tl_dirs_doing,
       $uniq_files,
       $files,
       $usedspc,
       $savedspc
      );
print "\nDONE\n";

__END__

=head1 NAME

trimtrees - traverse directories, find identical files, replace with hard links

=head1 SYNOPSIS

 trimtrees.pl directory...

=head1 DESCRIPTION

Traverse all directories named on the command line, compute MD5
checksums and find files with identical MD5. IF they are equal, do a
real comparison if they are equal and if so, replace the second of two
files with a hard link to the first one.

=cut



eg/thinnercache.pl
#!/usr/bin/perl -w

=pod

Schlage Perl Directories im /usr/local/ Directory zum Loeschen vor.
Based on Distance between neighbors. Born out of duenneresdone.pl

=cut

use strict;

use List::Util qw(reduce);
use File::Find;

use Getopt::Long;
our %Opt;
GetOptions(\%Opt, qw(max=i));

use Config;
my $max = 0;

my $root = shift or die "Usage: $0 directory";
die "Directory $root not found" unless -d $root;

my %no;
my %never_delete = map { $_ => undef }
    (
     1,41,170,273,379,496,   # this area needs manual intervention thus is precious
     266,267,     # Ilya's Jumbo patch
     9654,        # 5.6.1
     16904,16905, # Locale::TextDomain
     17637, # according to CoreList this is 5.008
     # 17705, # currently my oldest 5.9.0
     # 17706, # tentatively my oldest 5.9.0 with threads
     17656,17671, # JSON 2.07 (fixed in ticket 34599)
     17724,17725, # pseudo hashes
     17967,17968, # RCLAMP/Devel-LeakTrace-0.05.tar.gz
     18047,18048, # threads only: Devel::Caller
     18352, # 5.008
     ## 18408,18409, # PADBUSY for Scalar-Util-Clone-0.10

     # 19608, 19610, # wrong line number
     # 19391,19392, # maybe connected to Mail::Box, tests/nicholas-46463.pl
     19648,19649, # rt-perl-22977 ugly format bug and binary search needs help

     # 20556,20559, # DBI 1.58 (was a bug in perl)
     # 20939, # 5.008001 debugging
     21377, # according to CoreList this is 5.008001
     21670, # according to CoreList this is 5.008002
     21727, # 5.006001, Module::CoreList::patchlevel for 5.6.2 built with 'buildaperl --branch maint-5.6/perl-5.6.2 @21727'; installed as /home/src/perl/repoperls/installed-perls/maint-5.6/perl-5.6.2/p2dBQ7e/perl-5.6.1@21727/bin/perl

     ##22071,22074, # perlbug 53482 (magic/tie)

     22151, # according to CoreList this is 5.008003
     22314,22315, # B::Generate 1.06 without debugging, B::Generate 1.11 with debugging, Brackup
     # 22491, # 5.009001
     # 22552, # 5.008003 THREAD
     22729, # according to CoreList this is 5.008004
     ##22739,22741, # Devel::Profile 1.05
     22841,22842, # Encode::IMAPUTF7
     23139, # according to CoreList this is 5.008005
     23390, # 5.008005 THREAD
     23470,23471, # DCOPPIT/Mail-Mbox-MessageParser-1.5000.tar.gz
     23552, # according to CoreList this is 5.008006
     23767,23768, # Readonly::XS
     23963,23964, # Hatena::Keyword(not sure if I put this in the right line back then)
     23965,23966, # encoding neutral unpack: Data::Serializer 0.44 with JSON 2.07
     24009,24010, # Unicode::RecursiveDowngrade, Hatena::Keyword 0.05
     24448, # 5.008006 THREAD
     24517,24518, # removed AvFLAGS
     24521,24522, # removed HvPMROOT
     24525,24529, # 26 removed xhv_eiter from XPVHV and 29 regained compilability
     24541,24542, # Math::Pari
     24556,24557, # Math::Pari after rafl's patch
     24641, # according to CoreList this is 5.008007
     24659,24660, # Class::constr, Data::Structure::Util
     24965,24966, # Safe::World but only of marginal importance because failure has changed
     25178,25179, # CharsetDetector 1.0
     25567,25568, # srezic-rx-problem
     25805,25808, # Text::Query
     25965,25966, # Safe::World 0.14
     25985,25986, # Mail::Box
     26369,26370, # Class::MOP 0.40
     26373,26374, # classes
     26452,26453, # Devel::Profiler 0.04 "Not a glob reference"
     26454,26465, # Term::ReadPassword 0.07 (proxysubs; No useable patch between)
     26486,26487, # SQL::Translator; Class::MOP ?
     26876, # 5.008007 THREAD
     26979,26980, # Math::Pari after Nicholas' patch (Message-ID: <877ifnfoal.fsf@k75.linux.bogus>)
     27002, # 5.009003
     27040, # 5.008008 according to corelist
     27263,27264, # JSON 1.12
     27533,27534, # regexp recursion stuff/regression(rtperl#60034)
     27704,27705, # Thread.pm emulation: Net::Daemon
     28358,28359, # encoding::source
     28867,28868, # Regexp::Optimizer 0.15
     28966, # 5.009004
     29025,29026, # DBIx::Class(?), Authen::Htpasswd
     29317,29318, # String::Multibyte
     30038,30039, # forks 0.27
     30487,30488, # Devel::Cover 0.61
     30609,30610, # Sepia
     30637,30638, # XML-LibXML-Simple 0.12 (MarkOv) maint-5.8!
     30677,30678, # TAP::Parser 0.52
     30952,30953, # Thread.pm emulation: Net::Daemon
     30979,30980, # mro, Best 0.11, Class::Inner, Class::Prototyped
     31025,31027, # Regexp::Assemble
     31162, # 5.008008, my maint 5.8 installation under which I run Jifty
     31223, # 5.008008 THREAD
     31251,31252, # Data::Alias,Devel::EvalContext
     31940,31941, # Tcl (debugging perls)
     32147,       # my first bleadperl to run Jifty (yay!)
     32367,       # 5.10.0-RC1
     32491,       # 5.10.0-RC2
     32642,       # 5.10.0
     32706,32707, # breaks Data::Alias, Devel::Declare, autobox
     32733,32734, # DROLSKY/Class-MOP-0.65-70; replace RVs with IVs
     32739,32740, # breaks YAML::LibYAML due to ORANGE regexps
     32856,32857, # DCONWAY/Getopt-Euclid-v0.1.0.tar.gz know your enemy, this is hard to binarysearch, not the usual oneliner!
     33017,33018, # RCLAMP/Devel-Caller-2.03.tar.gz, Coro 4.37
     33021,33022, # DROLSKY/Devel-StackTrace-1.15.tar.gz, Devel-ebug-0.48
     33027,33028, # ROBIN/Want-0.16.tar.gz
     33029,33030, # ROBIN/PadWalker-1.6.tar.gz
     33056,33057, # ROBIN/Want-0.16.tar.gz again
     33071,33072, # PJCJ/Devel-Cover-0.63.tar.gz, ANDYA/Devel-LeakTrace-Fast-0.11.tar.gz
     33085,33086, # PAR::Packer 0.978
     33087,33088, # OLAF/Net-DNS-0.62.tar.gz
     33709,33710, # PJCJ/Devel-Cover-0.64.tar.gz
     33871,33874, # DLAND/Regexp-Assemble-0.33.tar.gz
     # current smoker
     # 24329, # 5.8.6
     26727, # 5.8.7
     27512, # 5.8.8
     34437, # 5.10.1 tobe
     34468, # 5.8.9 tobe
    );
ENDLESS: while () {
  my @dirs;
  find(sub {
         my $mname = $File::Find::name;
         $mname =~ s|\Q$root\E||;
         $File::Find::prune++ if $mname =~ tr[/][] > 2;
         return unless /perl.+\@(\d+)/;
         my $patch = $1;
         return if exists $never_delete{$patch};
         push @dirs, $File::Find::name if $mname;
       }, $root);
  my %n;

  # read them and give them a value
  warn "Found ". scalar @dirs ." deleteable perls";
  for my $dirent (@dirs) {
    next unless $dirent =~ m|/perl.+\@(\d+)|;
    next if exists $no{$dirent};
    $n{$dirent} = $1;
  }

  # sort them by value
  my @n = sort { $n{$a} <=> $n{$b} } keys %n;

  # we do not want to delete first and last one
  pop @n; shift @n;

  # group them by the difference to the next lower
  my %diff;
  for my $n (1..$#n) {
    my $diff = $n{$n[$n]} - $n{$n[$n-1]};
    $max = $diff if $diff > $max;
    $diff{$diff}{$n[$n]} = undef;
  }

  my $done;
  local($|) = 1;
 OUTER: for (my $i = 0; $i <= $max; $i++) {
    $done++ if $i == $max;
    next OUTER unless exists $diff{$i};
    # for my $d (sort { $n{$a} <=> $n{$b} } keys %{$diff{$i}}) {
  INNER: for my $d (reduce { $n{$a} < $n{$b} ? $a : $b } keys %{$diff{$i}}) {
      printf "unlink %s (distance %d) [Nyq]? ", $d, $i;
      my $ans;
      if ($Opt{max}) {
        sleep 1;
        if (@dirs > $Opt{max}) {
          $ans = "y";
        } else {
          $ans = "q";
        }
        print $ans, "\n";
        sleep 1;
      } else {
        $ans = <>;
      }
      if ($ans =~ /^y/i) {
        require File::Path;
        File::Path::rmtree($d) or die "Could not remove $d";
        if (-d $d) {
          die "ALERT: rmtree did not remove $d";
        } else {
          print "$d rmtreeed\n";
        }
      } elsif ($ans =~ /^q/i) {
        $done++;
      } else {
        print "Nothing done.\n";
        $no{$d} = undef;
      }
      last INNER;
    }
    last OUTER;
  }

  last ENDLESS if $done;
}


eg/relocate-perl.pl
#!/usr/bin/perl

=pod

Copy one of our perls in installed-perls to some place where we really
like to use it, say /usr/local. And correct every occurrence of the
miserable path we gave it within installed-perls to the new location.
Including, but not limited to the perl binaries and libperl.so and
whatever binary file is there.

I started out with

find /home/src/perl/repoperls/installed-perls/perl/pb0yHi3/perl-5.8.0@26561 -type f | xargs perl -nle 'if (m|/home/src/perl/repoperls/installed-perls/perl/pb0yHi3| &&!$seen{$ARGV}++) {printf "%d %s\n", -T $ARGV, substr($ARGV,45)}'

to determine which files I had to modify how to get a perl from one
path to another.

=cut

use strict;
use warnings;
use File::Rsync;
use File::Find;
use File::Path qw(mkpath);
use File::Spec;

sub Usage {
  "Usage: $0 from to";
}
my($from,$to) = @ARGV;
die Usage unless $to;
for ($from, $to) {
  s|/+$||;
}
die "to[$to] must be shorter than from[$from]" unless length($to) < length($from);
for ($from, $to) {
  die "dir[$_] not found" unless -e $_;
  die "dir[$_] not a directory" unless -d _;
  die "dir[$_] not absolute" unless File::Spec->file_name_is_absolute($_);
}
my $rsync = File::Rsync->new({ archive => 1});
$rsync->exec({src => "$from/", dest => "$to/"}) or die;
find(
     {
      wanted => sub {
        my $rel = substr($_,length($from));
        return unless $rel;
        return if -d $_;
        open my $fh, $File::Find::name or die "Could not open '$File::Find::name': $!";
        my $To = File::Spec->catfile($to,$rel);
        open my $tofh, ">", $To or die;
        if (-T $fh) {
          local $/ = "\n";
          while (<$fh>) {
            s/\Q$from\E/$to/g;
            print $tofh $_;
          }
        } else {
          local $/;
          local $_ = <$fh>;
          s/\Q$from\E([^\0]+)/ $to . $1 . ("\0"x(length($from)-length($to))) /ge;
          print $tofh $_;
        }
        close $fh;
        close $tofh;
      },
      no_chdir => 1,
     },
     $from
);


Makefile.PL
#!/usr/bin/perl -w -*- mode: cperl -*-
use 5.006001; # just being cautious, I did use "our" at least
use strict;
use ExtUtils::MakeMaker qw(:DEFAULT);
use File::Path;
eval { require File::Spec; };
my $HAVE_FILE_SPEC = !$@;
use vars qw($VERSION);

my $Id = q$Id: Makefile.PL 281 2008-01-20 21:04:15Z k $;
$VERSION = sprintf "%.3f", 1 + substr(q$Rev: 281 $,4)/1000;

my $version_diff = 0; # we'll have to die if this becomes true
my $version_from;
{
  local $^W;
  $ExtUtils::MakeMaker::VERSION = eval $ExtUtils::MakeMaker::VERSION;
  warn "Your MakeMaker is a bit dated[$ExtUtils::MakeMaker::VERSION].\nYou should get a new one\n"
        if $ExtUtils::MakeMaker::VERSION < 6.0;
}
if ($HAVE_FILE_SPEC) {
  $version_from = File::Spec->catfile(qw(lib Perl Repository APC.pm));
  my $version_set_manually = 0;

  if ($ARGV[0] && $ARGV[0] eq "--setversion") {
    die "Your perl is a bit dated[$]].\nDo not make a release with it\n" if $] < 5.008;
    die "Your MakeMaker is a bit dated[$ExtUtils::MakeMaker::VERSION].\nDo not make a release with it\n"
        if $ExtUtils::MakeMaker::VERSION < 6.3004;
    die "Your MakeMaker doesn't do the sign woodoo" unless
        MM->can("signature_target");
    shift @ARGV;
    my $st;
    $ENV{LANG} = "C";
    open $st, "svn st -u|" or die;
    my $dirty = 0;
    while (<$st>) {
      next if /^\?/;
      next if /^Head revision:/;
      next if /^Status against revision:/;
      print;
      $dirty++;
    }
    close $st;
    die "Not everything checked in or out?" if $dirty;

    # $re parses the upper boundary of the output of svnversion
    my $re;
    eval q{ $re = qr/(?:\d+:)?(\d+)[SM]*$/; };
    die $@ if $@; # old perl
    my($sv_dot)  = `svnversion .`             =~ $re;
    my($sv_from) = `svn st -u -v $version_from`  =~ /^.{8}\s*(?:\d+)\s*(\d+)/;
    die "Could not determine sv_from or sv_dot"
        unless $sv_from > 0 && $sv_dot > 0;
    if ($version_set_manually) {
      # we must control that the VERSION in CPAN.pm is the same as in the Makefile
      unshift @INC, "lib";
      require $version_from;
      open my $fh, "make the-release-name|" or die;
      my $have_version;
      while (<$fh>) {
        next unless /^version\s+([\d\._]+)/;
        $have_version = eval $1;
      }
      die "could not determine current version from Makefile" unless $have_version;
      eval q{
      no warnings "numeric";
      if ($CPAN::VERSION != $have_version) {
        warn "Not equal: CPAN::VERSION[$CPAN::VERSION] Makefile version[$have_version]";
        $version_diff = 1;
      }
};
      die $@ if $@;
    } else {
      if ($sv_dot > $sv_from) {
        warn "setting some property to make a checkin of this file possible";
        0==system(svn => "propset",
                  "cpan:release",
                  "after_$sv_dot",
                  $version_from) or die;
        warn "checking in to get the version up to the bleading edge";
        0==system(svn => "ci",
                  "-m",
                  "Property set for release after_$sv_dot",
                  $version_from) or die;
        # will have to die after Makefile is rewritten:
        $version_diff = 1;
      }
    }
    exit unless $version_diff;
  }
}

my @exefiles;
for my $f (
           "apc-overview",
           "binsearchaperl",
           "buildaperl",
           "patchaperlup",
           "perlpatch2svn",
           "apc2svn",
          ) {
  push @exefiles, File::Spec->catfile("scripts",$f);
}

my @sign = (MM->can("signature_target") ? (SIGN => 1) : ());
# warn "sign[@sign]";
WriteMakefile(
              NAME         => 'Perl::Repository::APC',
              VERSION_FROM => $version_from,
              EXE_FILES    => \@exefiles,
              PREREQ_PM    => {
                               "Digest::MD5"       => 0,
                               "File::Spec"        => 0,
                               "List::Util"        => 0,
                               "Module::CoreList"  => '2.13',
                               "Test::More"        => 0,
                               "version"           => '0.74',
                              },
              ($ExtUtils::MakeMaker::VERSION >= 6.3002 ?
               (LICENSE      => "perl") : (),
              ),
              @sign,
              dist => {
                         DIST_DEFAULT => 'setversion README all chlog tardist',
                         COMPRESS => 'gzip -9',
                        },
              EXTRA_META => <<EMETA,
repository: "https://franz.ak.mind.de/public-svn/Perl-Repository-APC"
EMETA
);

if ($version_diff){
  die "
==> I had to update some \$VERSIONs <==
==> Your Makefile has been rebuilt. <==
==> Please rerun the make command.  <==
";
}

package MY;

sub postamble {
  q{setversion:
	$(PERL) Makefile.PL --setversion

chlog :: chlog-svn chlog-trad

chlog-svn :
	svn log -v https://franz.ak.mind.de/public-svn/Perl-Repository-APC > ChangeLog.svn

# got this from http://ch.tudelft.nl/~arthur/svn2cl/
chlog-trad :
	-/home/src/devel/svn/svn2cl/svn2cl-current/svn2cl.sh --strip-prefix trunk --group-by-day https://franz.ak.mind.de/public-svn/Perl-Repository-APC > ChangeLog

the-release-name :
	$(NOECHO) $(ECHO) 'version ' $(VERSION)
	$(NOECHO) $(ECHO) 'release-name ' $(DISTVNAME).tar$(SUFFIX)

release ::
	/usr/bin/perl -I../../CPAN/SVN/release-lib -S release $(DISTVNAME).tar$(SUFFIX)
	rm -rf $(DISTVNAME)

}
}

sub libscan {
    my $path = $_[1];
    return '' if $path =~ /\B\.svn\b/;
    return $path;
}

sub dist_ci {
  return qq{ci :
	svn ci
};
}

# I want to keep the fresh signature here and not hidden in the
# tarball
sub dist_test {
  return q{
# if we depend on $(DISTVNAME).tar$(SUFFIX), then the rest of the
# Makefile breaks our intent to NOT remake dist
disttest :
	rm -rf $(DISTVNAME)
	tar xvzf $(DISTVNAME).tar$(SUFFIX)
	cd $(DISTVNAME) && $(ABSPERLRUN) Makefile.PL
	cd $(DISTVNAME) && $(MAKE) $(PASTHRU)
	cd $(DISTVNAME) && $(MAKE) test $(PASTHRU)

distdir ::
	$(CP) $(DISTVNAME)/SIGNATURE ./SIGNATURE
	$(CP) $(DISTVNAME)/META.yml  ./META.yml
	$(CP) $(DISTVNAME)/MANIFEST  ./MANIFEST

}
}

sub distdir {
  my $self = shift;
  my $out = $self->SUPER::distdir;
  $out =~ s/distdir :/distdir ::/g;
  return $out;
}

__END__


=head1 Valuable advice I got

http://rgarciasuarez.free.fr/perl/svnmm.html

=head1 In development

Currently I'm developing the mechanism for a convenient 'make dist'.

1. The first thing that 'make dist' has to do is making sure that
   $VERSION is right in at least one module, the VERSION_FROM module.

2. We add a target called 'setversion' to DIST_DEFAULT:

    DIST_DEFAULT => 'setversion all tardist',

3. We define the target in a postamble:

    sub MY::postamble {
      q{setversion:
            $(PERL) Makefile.PL --setversion
    }
    }

4. When the Makefile is called with --setversion, we do the following:

4.1. We verify that everything has been checked in. We allow only '?'
     in the first column of the output of 'svn st -u'. If we have
     something else, we die, halting any pending make targets.

4.2. We compare what svnversion has to say about "." and the
     VERSION_FROM module. If the upper limit of "." has a higher rev
     than the VERSION_FROM module, then we set a "cpan:release"
     property on the VERSION_FROM module and check it in. Otherwise we
     exit immediately and are done.

4.3. Then we write a new Makefile

4.4. And finally we die, so that the calling make also dies. Our
     message says, "...please rerun the make command".

5. As 'make dist' now always starts with the 'setversion' target, we
   call 'Makefile.PL --setversion' early and fail if we either have
   not checked in or our VERSION_FROM module isn't fresh enough.

6. On an immediate second call to 'make dist' the VERSION_FROM module
   is always fresh enough and the Makefile reflects the exact $VERSION
   we have in the VERSION_FROM module.


=cut


